include/system/analytics_manager.js

```diff
@@ -108,11 +136,184 @@ module.exports = function AnalyticsManagerModule(pb) {
+            });
+            d.once('error', AnalyticsManager.buildDomainErrorHandler(opts));
+        };
+    };
+
+    /**
+     *
+     * @static
+     * @method buildProviderResponseHandler
+     * @param {Object} opts
+     * @param {String} opts.key
+     * @param {String} opts.method
+     * @param {String} opts.url
+     * @param {Integer} opts.start The epoch when the provider started execution
+     * @param {Object} opts.th The timeout handle for the provider execution
+     * @param {Function} opts.callback
+     * @return {Function}
+     */
+    AnalyticsManager.buildProviderResponseHandler = function(opts) {
+        return function(err, result) {
+            if (util.isError(err)) {
+                pb.log.error("AnalyticsManager: Rendering provider [%s] failed for URL [%s:%s]\n%s", opts.key, opts.method, opts.url, err.stack);
+            }
+
+            //collect runtime metrics
+            AnalyticsManager.updateStats(opts.start, opts.key);
+
+            //we only callback if the timeout handle exists.  When
+            //null, it means that the task already timed out and we
+            //shouldn't attempt to send back a result.
+            if (opts.th !== null) {
+                clearTimeout(opts.th);
+                opts.th = null;
+
+                //the error is left out on purpose.  It is logged above
+                //and we want all other providers to have a chance.
+                opts.callback(null, result);
+            }
+        };
+    };
+
+    /**
+     *
+     * @static
+     * @method buildDomainErrorHandler
+     * @param {Object} opts
+     * @param {String} opts.key
+     * @param {String} opts.method
+     * @param {String} opts.url
+     * @param {Integer} opts.start The epoch when the provider started execution
+     * @param {Object} opts.th The timeout handle for the provider execution
+     * @param {Function} opts.callback
+     * @return {Function}
+     */
+    AnalyticsManager.buildDomainErrorHandler = function(opts) {
+        return function(err) {
+            pb.log.error('AnalyticsManager: Rendering provider [%s] failed for URL [%s:%s]\n%s', opts.key, opts.method, opts.url, err.stack);
+
+            //only callback if the timeout handle is still in place
+            if (opts.th !== null) {
+                clearTimeout(opts.th);
+                opts.th = null;
+
+                //collect runtime metrics
+                AnalyticsManager.updateStats(opts.start, opts.key);
+                opts.callback(null, DEFAULT_RESULT);
+            }
+        };
+    };
+
+    /**
+     * Updates the runtime statistics to include
+     * @static
+     * @method getStats
+     * @return {Object}
+     */
+    AnalyticsManager.updateStats = function(startTime, provider) {
+        var runTime = (new Date().getTime()) - startTime;
+        updateStatBlock(STATS.cumulative, runTime);
+        updateStatBlock(STATS.providers[provider], runTime);
+    };
```

Comment:
As part of performance tracking we now keep stats on the average runtime for the providers (except when it errors).  It tracks individual providers (not on a site basis) and overall in the event that there are multiple providers